CONTRIBUTING.md

```diff
@@ -0,0 +1,142 @@
+
+
+# Error handling
+
+How to handle errors in Ansible.
+
+## Tracebacks
+
+Do not manually generate tracebacks for errors or warnings. Standardized traceback capture is included for errors, warnings and deprecation warnings for both controller and module code (Python). See the `DISPLAY_TRACEBACK` config option for details on how to enable traceback capture and display.
+
+## In modules
+
+In most cases, just raise an exception. The AnsiballZ wrapper now provides a general exception handler for Python modules, making use of the `AnsibleModule.fail_json` method unnecessary, unless the module result needs to be customized. Calls to the various Python-module-side `warn` and `deprecate` methods/functions will also capture and marshal tracebacks to the controller when enabled.
+
+## Deferred exceptions in modules
+
+When deferring exceptions in modules using `try/except` and `fail_json`, pass the captured `Exception` instance to `fail_json` using the `exception` argument. The error handling infrastructure will handle collection of error details and traceback formatting.
+
+## Exception context
+
+Raising exceptions while another exception is active will cause the active exception to become the `__context__` for the newly raised exception. This is usually not the intended behavior, and should be reserved for unexpected errors while handling the original exception. In most cases a `raise from` is desired, either with `None` to suppress the original exception if it's not helpful, or with the captured exception to set it as the `__cause__` for the newly raised exception.
+
+In earlier versions of Ansible, this could be done by setting `orig_exc` on `AnsibleError`. However, this is no longer needed and a simple `raise from` should be used instead.
+
+Suppression example: `raise Exception("something") from None`
+Cause example: `raise Exception("something") from ex`
+
+## Raising new exceptions
+
+Don't catch exceptions just to re-raise them, unless there's additional information that can be added in the newly raised exception. In most cases, particularly for plugin/module failures, contextual information is automatically added, making granular `try/except/raise` within the plugin or module unnecessary.
+
+## Error messages
+
+Do not repeat previous exception messages when constructing new exceptions.
+
+Anti-pattern: `raise Exception("it broke: {ex}") from ex`
+
+The built in error chain handling mechanisms in Ansible will include the messages from cause/context exceptions automatically. It will also make simple attempts at de-duplicating messages to compensate for existing code which goes against this recommendation.
+
+In general, error messages passed to exceptions should be a fairly terse description of what happened, and not contain extra diagnostic, contextual, or prescriptive correction advice (see `obj` and `help_text` below for more on that).
```

Comment:
```suggestion
Do not repeat previous exception messages when constructing new exceptions.

Anti-pattern: `raise Exception("it broke: {ex}") from ex`

The built-in error chain handling mechanisms in Ansible will include the messages from cause/context exceptions automatically. It will also make simple attempts at de-duplicating messages to compensate for existing code, which goes against this recommendation.

In general, error messages passed to exceptions should be a fairly terse description of what happened, and not contain extra diagnostic, contextual, or prescriptive correction advice (see `obj` and `help_text` below for more on that).
```